### ISSUE TITLE
"say $obj.^methods" does not list all methods

### ISSUE THREAD

**Issue by doomvox** [rakudo/rakudo#4207]
## The Problem

The "^methods" method, when used in the obvious way does not
always give you a list of all available methods.  The method
"^methods" returns Method objects which don't stringify correctly
when roles are involved.  

## Expected Behavior

When used in the obvious way "^methods" should yield a list
of all available methods.

## Actual Behavior

On more complex objects/classes, when used in the obvious way,
"^methods" returns only a partial list of methods, concealing
some behind the peculiar value "Method+{is-nodal}.new".

## Steps to Reproduce

Run "^methods" on a Set, dump the result with "say":

    my $s = set 2, 4, 6;
    say $s.^methods;

       (iterator of new-from-pairs default elems total grab grabpairs pick pickpairs roll keyof STORE Int Num Numeric Real Capture fmt Setty Baggy Mixy WHICH Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new new Method+{is-nodal}.new Method+{is-nodal}.new minpairs maxpairs Bool Method+{is-nodal}.new Method+{is-nodal}.new ACCEPTS Str gist raku Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new Method+{is-nodal}.new BUILDALL)

Note: there are a number of entires in this list that don't make any
sense at first glance:

    "Method+{is-nodal}.new"

Also note: there are methods that exist with Sets that aren't
included in this list, such as "keys" and "values".

To actually get a list of all methods, you need to know about the
"name" method and apply it to each Method object yourself:

     $s.^methods.map( *.name ).say;

     (menu Capture Baggy DELETE-KEY maxpairs hash default STORE total Numeric SET-SELF grabpairs Bool fmt RAW-HASH minpairs elems pickpairs Bag SetHash ASSIGN-KEY pairs new values keyof kv Setty EXISTS-KEY WHICH Str of MixHash Int gist Mix Num list raku pick grab iterator Real Set Mixy AT-KEY roll Hash ACCEPTS antipairs new-from-pairs keys BagHash BUILDALL)

Note: that looks like a neater, more comprehensive list of
methods, including "keys" and "values".

Either Method objects should be improved so that .gist returns
the same as .name; or (possibly) the documentation should be
improved to explain how to use "^methods" correctly,
i.e. "^methods.map(*.name).say".

Note that "^methods" is not at all obscure: it's mentioned on p.3
of brian d foy's "Learning Perl6".

Original discussion:

https://www.nntp.perl.org/group/perl.perl6.users/2021/02/msg9734.html

## Environment
* Operating system:

  uname -a
  Linux fandango 4.9.0-8-686 #1 SMP Debian 4.9.144-3 (2019-02-02) i686 GNU/Linux

* Compiler version (`perl6 -v` or `raku -v`):

  raku -v
  Welcome to 𝐑𝐚𝐤𝐮𝐝𝐨™ v2020.10.
  Implementing the 𝐑𝐚𝐤𝐮™ programming language v6.d.
  Built on MoarVM version 2020.10.

**Comment by ugexe** [comment]
What about people that want to access the actual method objects themselves? And those that already do so with this syntax? There is more to introspection than making things easy for a human to read from a terminal after all.

**Comment by fecundf** [comment]
> What about people that want to access the actual method objects themselves? And those that already do so with this syntax? 

The answer is -
> Method objects should be improved so that .gist returns the same as .name

Thus no change to what ^methods returns, rather stringifying a method via .gist should delegate to .name

**Comment by ugexe** [comment]
> Thus no change to what ^methods returns, rather stringifying a method via .gist should delegate to .name

ah yeah I misparsed that sentence when I read it; changing the gist method seems entirely reasonable

**Comment by MasterDuke17** [comment]
It looks like it already does?
https://github.com/rakudo/rakudo/blob/master/src/core.c/Method.pm6#L4

But maybe
https://github.com/rakudo/rakudo/blob/master/src/core.c/core_prologue.pm6#L51
needs to happen in the epilogue instead?

**Comment by jubilatious1** [comment]
I started analyzing results posted in the <perl6-users@perl.org> mailing list, but was unsure if there was a change AFTER the Rakudo.2020.10 release. Below is my 2020.10 result (50 elements):

```
user@mbook:~$ raku
Welcome to 𝐑𝐚𝐤𝐮𝐝𝐨™ v2020.10.
Implementing the 𝐑𝐚𝐤𝐮™ programming language v6.d.
Built on MoarVM version 2020.10.

To exit type 'exit' or '^D'
> my $s = set 2, 4, 6;
Set(2 4 6)
> put $s.^methods.elems;
50
> dd $_ for $s.^methods>>.name.sort;
"ACCEPTS"
"ASSIGN-KEY"
"AT-KEY"
"BUILDALL"
"Bag"
"BagHash"
"Baggy"
"Bool"
"Capture"
"DELETE-KEY"
"EXISTS-KEY"
"Hash"
"Int"
"Mix"
"MixHash"
"Mixy"
"Num"
"Numeric"
"Real"
"STORE"
"Set"
"SetHash"
"Setty"
"Str"
"WHICH"
"antipairs"
"default"
"elems"
"fmt"
"gist"
"grab"
"grabpairs"
"hash"
"iterator"
"keyof"
"keys"
"kv"
"list"
"maxpairs"
"minpairs"
"new"
"new-from-pairs"
"of"
"pairs"
"pick"
"pickpairs"
"raku"
"roll"
"total"
"values"
Nil
>
```

CC: @vrurg

**Comment by doomvox** [comment]
jubilatious1 <notifications@github.com> wrote:

>> my $s = set 2, 4, 6;
> Set(2 4 6)
>> put $s.^methods.elems;
> 50
>> dd $_ for $s.^methods>>.name.sort;

Yes, that's right, if you use .name you get method names.

But if you use .gist:

   dd $_ for $s.^methods>>.gist.sort;

You'll get a bunch of method names, but some from the previous list
will be missing, and in their place you'll see:

   "Method+\{is-nodal}.new"

And if you do something naive just using say, you get an implicit use
of .gist, and end up with a similar problem:

    say $s.^methods;

I haven't double-checked but I've got a strong impression that the
trouble is with methods tthat were brought in via Roles as opposed to
the ones inherited from classes.

**Comment by jubilatious1** [comment]
Yes, that seems confusing.

**Comment by jonathanstowe** [comment]
> I haven't double-checked but I've got a strong impression that the trouble is with methods tthat were brought in via Roles as opposed to the ones inherited from classes.

They may be methods to which a role has been *applied* which may have happened as part of the application of a trait or by some other means. 

I'm somewhat surprised that it hasn't been mentioned that `.^methods` may not be at all suitable for introspecting the names of _usable_ methods (whether using `.name` or the `gist`,) as you will get the name that was defined for the Method object (or even `<anon>` and _not the name by which you would call the method on the type_.

See for example:

```
raku -e 'class F {}; F.^add_method("blub", (my method foo() { say "foo" }));  say F.^methods; F.foo;'
(BUILDALL foo)
No such method 'foo' for invocant of type 'F'.  Did you mean 'foo'?
  in block <unit> at -e line 1

```
vs

```
raku -e 'class F {}; F.^add_method("blub", (my method foo() { say "foo" }));  say F.^methods; F.blub;'
(BUILDALL foo)
foo
```

Infact it is entirely possible that several of the Methods reported by `.^methods` may have the same `name` but supply a different implementation being a closure, e.g.:

```
class F {
}

for <blub baz bar> -> $n {
    F.^add_method($n, my method foo() { say $n });
}

say F.^methods;
```

Of course in reality it is more likely that in the above case the method would be declared without a name and thus be reported as `<anon>`, but this is illustrative of the point that the name of the Method object is not that of the resulting class method.

The actual names of the methods callable on the class are available as `.^method_names` and the Hash returned by `.^method_table`  relates the usable name to the underlying Method object so to elaborate the above example:

```
class F {
}

for <blub baz bar> -> $n {
    F.^add_method($n, my method foo() { say $n });
}

for F.^method_table.kv -> $name, $method {
    say "$name : { $method.name }";
};
```

It makes clearer what is going on.  (One slight caveat on this, I don't know whether the above are specced, considered an implementation detail or subject to peremptory removal.)

So in summary I don't think there is an actual problem here beyond the clarity of the documentation and the possible inaccuracy of certain "folk knowledge".

**Comment by doomvox** [comment]
There *certainly* are some issues with documentation. There's no
mention whatsover of any ^method* methods here, for example:

  https://docs.raku.org/language/mop#index-entry-Introspection

However, the ^methods method is certainly not obscure-- as I
mentioned, I've seen it in brian d foy's "Learning Perl",
and that's not the only place.    (And as an aside; it's severely
painful to try to searches on this, the terms are used everywhere.)

If ^method_names is the Right Way to do this, I'm okay with that,
but hardly anyone seems to know about it, which is not okay.

Is there any particularly reason that the behavior of

    $s.^methods>>.gist

Shouldn't be modified to bring it more into line with what
^method_names  does?

**Comment by jonathanstowe** [comment]
Yes there is a reason that it *can't* be modified thus, which I hoped would be made clear by the example above.  The point being that the Method objects returned by `^methods` have no need to know into which classes and by which actual names they have been composed, the same Method could be composed into multiple classes by different names for instance.

Unfortunately any place that suggests that `^methods` is a good way of determining the names of methods that can be called on a class is basically wrong in that regard, basically faulty "folk knowledge" that is neither backed up by the spec tests that define Raku nor the documentation in https://docs.raku.org/type/Metamodel::MethodContainer (which admittedly could make this a tad clearer.)

I'd suggest that we take this to a discussion on https://github.com/Raku/doc to clarify this and someone can suggest to brian that he change that in a subsequent edition of his book.

**Comment by vrurg** [comment]
> I'm somewhat surprised that it hasn't been mentioned that `.^methods` may not be at all suitable for introspecting the names of _usable_ methods (whether using `.name` or the `gist`,) as you will get the name that was defined for the Method object (or even `<anon>` and _not the name by which you would call the method on the type_.

You certainly haven't followed the discussion closely. `.name` will always give you the right method name. There is something wrong about `.gist` though.

Otherwise, `.^methods` clearly does what it supposed to do: it gives a list of `Method` objects. What one does with them is totally different story.

**Comment by jonathanstowe** [comment]
Au contraire my buddy I have been following it perfectly, including the subtext, there may be something wrong with the `gist` of some Method objects but that isn't the fault of Method which defines it as `method gist { self.name }` but possibly of some role or subclass of Method which may define it differently.  However it is evident that @jubilatious1 and @doomvox have been led to believe that the "names" (or gist )  of the Method objects are the names that may be used to call those methods on a given class, which is fundamentally incorrect as I tried to demonstrate by example above.  But yes `^methods` is working exactly as it is designed, there is just some "fake news" about what that means.

**Comment by doomvox** [comment]
Starting with the list of types from the documentation:

   https://docs.raku.org/type.html

And trying to call ^methods on all of them, I find 48 of what I
would call problem cases, where calling .gist on each method
object returns the string 'Method+{is-nodal}.new'.

Here's a list of the problem cases, including a count of the
number of problematic methods on each:

     [1 Channel]
     [1 IO::Handle]
     [1 IO::Pipe]
     [1 IO::Socket::Async]
     [1 Promise]
     [2 IO::ArgFiles]
     [2 IO::CatHandle]
     [2 NFC]
     [2 NFD]
     [2 NFKC]
     [2 NFKD]
     [2 Uni]
     [3 Endian]
     [3 Order]
     [4 Backtrace]
     [4 Nil]
     [4 utf8]
     [6 IO::Path::Parts]
     [8 Pair]
     [8 Supply]
     [9 Capture]
     [9 Grammar]
     [9 Match]
     [14 HyperSeq]
     [14 RaceSeq]
     [16 Seq]
     [17 Range]
     [17 SetHash]
     [18 Set]
     [19 Bag]
     [19 BagHash]
     [19 Mix]
     [19 MixHash]
     [20 Map]
     [22 PseudoStash]
     [25 Hash]
     [26 Stash]
     [36 List]
     [38 Slip]
     [51 Array]
     [75 Any]
     [75 Metamodel::ClassHOW]
     [75 Metamodel::ConcreteRoleHOW]
     [75 Metamodel::CurriedRoleHOW]
     [75 Metamodel::EnumHOW]
     [75 Metamodel::PackageHOW]
     [75 Metamodel::ParametricRoleGroupHOW]
     [75 Metamodel::ParametricRoleHOW]

**Comment by doomvox** [comment]
> there may be something wrong with the gist of some Method objects but that isn't the fault of Method which defines it as method gist { self.name }

My apologies if I'm unfairly maligning the Method class.  The main point here is I'm trying to report what seems like puzzling behavior at the user level.   It doesn't seem like there can be anything useful about having .gist return a string like 'Method+{is-nodal}.new', not when there really is an obviously valid .name.    Any suggestions on tracking down the source of this puzzling behavior would be appreciated. 

You're making a lot of interesting points about the way things can behave, but I'm not talking about any unusual corner cases, I'm talking about fairly straight-forward  features like the values and keys methods on QuantHash objects.

**Comment by doomvox** [comment]
Going back to the design documents, "say()" is specified to use ".gist":

   https://design.raku.org/S32/IO.html

   Before printing, call the .gist method on any non-Str objects.

And "gist" is described here:

  https://design.raku.org/S99.html#gist

  A summary, or a method on objects that is invoked by "say" to
  print out the object in a human readable format.

I submit that 'Method+{is-nodal}.new' does not count as "human
readable", and the behavior I'm describing here is an actual bug by
any reasonable standard.

About that idiom of looping over method objects and manually
calling .name:

   say .name for Set.^methods();

I think this is essentially a workaround for the bug that many people
have gotten used to and now think of as normal.

This is more natural, and the thing a new user is likely to try:

  say Set.^methods();

Further there's nothing in the spec or the design documents that
suggest that the need to manually call .name is correct behavior.

There's also nothing anywhere in the specs or the docs about
"^method_names".

**Comment by doomvox** [comment]
"^methods" is described in the design documents:

https://design.raku.org/S12.html#Introspection

     MyClass.^methods()          # get the method list of MyClass

     The .^methods method returns method-descriptors containing:

         name                the name of the method

I don't see any mention of the issue that Jonathan Stowe has raised,
that there might hypothetically be a difference between the name of
the method where it was definied and the name you would call
it by now.  (This *is* a hypothetical case, isn't it?  I've yet 
to see something like that in actual code.)

The general flavor of Synopsis 12 suggests that "^methods" is
expected to be used by users, it's not just for internal use:

  "The motivation for stopping at Cool and Any by default is
  that the first derivation from one of those is really what
  the user thinks of as the root class of the type hierarchy
  [...]  When the user asks for $object.^methods, listing such
  methods by default is rather irritating."

There are some roast tests for "^methods", but those are
mostly sanity checks on the numbers of method objects
returned by it.  I don't think the behavior of calling gist
on a method object is examined anywhere in roast.

**Comment by jonathanstowe** [comment]
> This _is_ a hypothetical case, isn't it? I've yet
> to see something like that in actual code.

Nope,

```
[jonathan@menenius katikati]$ cat mm && raku mm
class Foo {

    role Z {
        method gist {
            'whatever';
        }
    }

    multi sub trait_mod:<is> ( Method $m, :$z! ) {
        $m does Z;
    }

    method vlub is z {
    }

    submethod TWEAK {
        self.^add_method('blah', method () {});
    }
}


say Foo.new.^methods;

# vim: ft=raku
(whatever TWEAK Submethod+{is-hidden-from-backtrace}.new <anon>)
```

You need to use `^method_names` if you want the names of the callable methods regardless of what the synopses may say.

**Comment by doomvox** [comment]
Thanks, that's a very clear example that shows a case where
"^method_names" does a better job than the competition:

    my $foo = Foo.new;

    say $foo.^methods;
    ## (whatever TWEAK Submethod+{is-hidden-from-backtrace}.new <anon>)

    say $foo.^methods>>.name;
    ## (vlub TWEAK BUILDALL <anon>)

    say $foo.^method_names;
    ## (vlub TWEAK BUILDALL blah)

But even this example shows a case where invoking .name
manually outperforms the implicit .gist call: .name says 
"BUILDALL" where .gist just says:

    Submethod+{is-hidden-from-backtrace}.new

So something out there has a broken .gist, because that
doesn't seem to be "human readable".

In the case of "blah", ^method_names is indeed doing a better
job of finding the name of a callable method (just labeling
it "<anon>" is wrong: it's not anonymous any more).  I would
speculate though that if ^method_names can do this, the
Method's gist routine could be improved to do the same job.

And the "blah" case you've demoed here really doesn't have
a lot of resemblence, to the case I've been talking about,
where .gist fails but .name succeeds:

    say Set.^methods>>.name.grep(/values/)
    ## (values)
    say Set.^methods>>.gist.grep(/values/)
    ## ()

Again, when there's a .gist that doesn't give you something
"human readable" that would seem to be a bug.

It's nice to know that this works:
 
    say Set.^method_names.grep(/values/)
    ## (values)

But it's never been mentioned in specs, synopses or
documentation.

**Comment by jubilatious1** [comment]
Not clear to me. Try `^method`, try `^methods`, try `^methods_names`, try `^method_names`, in various combinations:

```
===03/04/21 14:16:04===
(whatever TWEAK Submethod+{is-hidden-from-backtrace}.new <anon>)
===03/04/21 14:16:43===
(whatever TWEAK Submethod+{is-hidden-from-backtrace}.new <anon>)
Package 'Foo' already has a method 'blah' (did you mean to declare a multi-method?)
  in submethod TWEAK at - line 17
  in block <unit> at - line 23

===03/04/21 14:17:21===
No such method 'methods_names' for invocant of type
'Perl6::Metamodel::ClassHOW'.  Did you mean 'method_names'?
  in block <unit> at - line 23

===03/04/21 14:17:44===
(vlub TWEAK BUILDALL blah)
===03/04/21 14:18:23===
No such method 'method' for invocant of type
'Perl6::Metamodel::ClassHOW'.  Did you mean 'methods'?
  in block <unit> at - line 22

```

**Comment by doomvox** [comment]
jubilatious1: the code example that Jonathan gave us breaks when you try to call Foo.new a second time.  That's why when I used it I did it just once and then used the returned object:  

     my $foo = Foo.new;
     say $foo.^methods>>.name;
     say $foo.^method_names;

Presumably if you really wanted to use this add_method trick in real code, you'd do something more complicated, but it doesn't seem like the main issue here, so I just worked around it.

**Comment by jubilatious1** [comment]
Hi @doomvox I was just trying to illustrate how it might be difficult for a newbie to stumble upon the correct incantation with respect to the family of `method` keywords. Each time/date such as `===03/04/21 14:17:44===` represents a different run of @jonathanstowe 's code. 

I'd say that we might have a useful and memorable idiom here if it's common-usage to 1. take the plural function name (e.g. `methods`), 2. drop the terminal `s`, and 3. add the `_names` (plural) as a suffix giving `method_names`. Searching the docs I see that the methods `named_names` and `private_method_names` _might_ fall into this category, but it's not an idiom I've investigated to any degree (and that makes only three extant instances...).

**Comment by doomvox** [comment]
I see what you were getting at.  If you're asking yourself "how can I get a list of methods?"  what would you try? 

My point though is that I didn't just guess, I looked up what I was supposed to do, and what you're told only sort-of works, but only sometimes, and when it does weird things it's hard for a beginner to figure what's going on.

**Comment by mykhal** [comment]
Just wanna say I'm resonating, found this thread after my surprise that
```raku
"a".^methods>>.gist.Set !(==) "a".^method_names.Set
```
Namely,

I did not find `"uniname"` in `"a".^methods>>.gist.Set`,  
then also not in `Str.^methods>>.gist.Set`,
then also not in `Uni.^methods>>.gist.Set`,
nor `Cool.^methods>>.gist.Set`,
but then surprisingly did in `"a".^method_names`.

```
Welcome to 𝐑𝐚𝐤𝐮𝐝𝐨™ v2021.06.
Implementing the 𝐑𝐚𝐤𝐮™ programming language v6.d.
Built on MoarVM version 2021.06.
To exit type 'exit' or '^D'
> "a".^method_names (-) "a".^methods>>.gist ;
Set(BUILDALL WORDS_AUTODEREF naive-word-wrapper ord ords unimatch uniname uninames uniparse uniprop uniprops unival univals)
```

**Comment by MasterDuke17** [comment]
@Kaiepi you've done a lot with roles, think this is something you could easily fix?

**Comment by Kaiepi** [comment]
@MasterDuke17 there's a tedious, but easy fix for cases of this problem that pop up in core, but not so much the root of it all. The methods lacking the proper `gist` because of traits performing mixins used to not typecheck like `Callable` either. A similar fix for [that](https://github.com/rakudo/rakudo/commit/7f2ae26cdfda0a598a7b3c9fa513000553490640) works for this:
```
bastille% raku -e '.^compose for Set.^methods; say Set.^methods'
(iterator of new-from-pairs default elems total Map grab grabpairs pickpairs roll keyof item STORE Int Num Numeric Real Capture fmt Setty Baggy Mixy WHICH kv values Set SetHash AT-KEY ASSIGN-KEY DELETE-KEY new keys antipairs minpairs maxpairs Bool hash Hash ACCEPTS Str gist raku pick EXISTS-KEY Bag BagHash Mix MixHash list pairs BUILDALL)
```
But I think there's a more general problem of how to deal with composition of mixins when their target isn't yet. In chicken-and-egg scenarios like this, they shouldn't demand ad-hoc augmentation to behave.

**Comment by doomvox** [comment]
This file exercises .^methods more thoroughly than existing roast tests:

  https://github.com/doomvox/darkroast/blob/main/S12-introspection/method_object_gists.t

Many of the tests fail when run with an older release of raku:

  Welcome to Rakudo(tm) v2021.04.

But it passes when run with a more recent build:

  Welcome to Rakudo™ v2022.12-1-gd52342eb0.

So, I'm closing this out as fixed.  Many thanks to the dev team.

**Comment by doomvox** [comment]
As mentioned, this works for recent builds of raku.  Definitely fixed.
